garnett/ciffilereader.py

```diff
@@ -193,13 +208,17 @@ def read(self):
             unique_points = np.array(unique_points, dtype=np.float32)
 
             if bad_types:
+                site_types = [self.default_type]
                 type_strings = [self.default_type] * len(unique_points)
         else:
             unique_points = fractions
             type_strings = site_types
 
-        # Convert type strings to types, typeid
-        types, typeid = _generate_types_typeid(type_strings)
+        # Remove duplicate site_types while preserving their order
+        site_types = list(OrderedDict.fromkeys(site_types))
+
+        # Convert type strings to typeid
+        typeid = [site_types.index(t) for t in type_strings]
```

Comment:
It probably won't be the end of the world if we don't do this, but could you add a simple dict for the `type_name -> typeid` map and use that here?

Reply:
```suggestion
        site_types_lookup = {key: index for index, key in enumerate(site_types)}
        typeid = [site_types_lookup[t] for t in type_strings]
```